EZP-24699: Enrich policy limitations view with human readable names

adamwojs opened the pull request:
> JIRA: https://jira.ez.no/browse/EZP-24699

## Description

_Disclaimer: There wasn't significant activity in original JIRA ticket since 2016, so I decided to create this  PR as my own open-source contribution_ :wink:

This PR contains implementation of mapping the policy limitation values to human readable versions. 

![zrzut ekranu z 2017-09-25 22-43-47](https://user-images.githubusercontent.com/211967/30830084-0adaea66-a243-11e7-8914-6fda9faa7c6c.png)

## Architecture

Limitation values are rendered via `ez_render_limitation_value` Twig function, which takes two arguments: 

1. `eZ\Publish\API\Repository\Values\User\Limitation` 
2. Optional hash of parameters

and delegates execution to `EzSystems\RepositoryForms\Templating\LimitationBlockRenderer`. 

Limitation values are mapped by services implementing the following interface: 

```php
/**
 * Interface for Limitation Value mappers.
 */
interface LimitationValueMapperInterface
{
    /**
     * Map the limitation values, in order to pass them as context of limitation value rendering.
     *
     * @param Limitation $limitation
     * @return mixed[]
     */
    public function mapLimitationValue(Limitation $limitation);
}
```
and registered in DIC with `ez.limitation.valueMapper` tag with `limitationType` attribute e.g.

```yml
ezrepoforms.limitation.form_mapper.newsection:
    parent: ezrepoforms.limitation.form_mapper.multiple_selection
    class: "%ezrepoforms.limitation.form_mapper.section.class%"
    arguments: ["@ezpublish.api.service.section"]
    tags:
        - { name: ez.limitation.formMapper, limitationType: NewSection }
        - { name: ez.limitation.valueMapper, limitationType: NewSection }
```

Responsibility of `EzSystems\RepositoryForms\Templating\LimitationRenderer` is similar to `\eZ\Publish\Core\MVC\Symfony\Templating\Twig\FieldBlockRenderer`: render propery twig block and supply all required parameters: 

* limitation instance
* mapped values 
* parameters (second argument of `ez_render_limitation_value`) 

Block name is resolved as `ez_limitation_%LIMITATION_IDENTIFIER%_value`
, where `%LIMITATION_IDENTIFIER%` is identifier of limitation (converted to lower case characters). 

If the value mapper for limitation type is not provided, the "fallback block" is rendered (`ez_limitation_value_fallback`).

## TODO:

- [X] Implementation
- [x] Tests
- [ ] Documentation for `ez_render_limitation_value` and `eZ\Publish\API\Repository\Values\User\Limitation\LimitationValueMapperInterface`

lserwatka commented:
Thank you @adamwojs for this PR. It brings a lot value.

webhdx commented:
Looks good so far. Great contribution.

andrerom commented:
Looks overall good, added @glye as reviewer as he was involved in making this and the story around for the names.

adamwojs commented:
Thank you for yours feedback. 

> General question: what do I need to make it work in PlatformUI?

@alongosz Its need change of few lines in PlatformUI: https://github.com/ezsystems/PlatformUIBundle/pull/907 :wink:

glye commented on lib/Limitation/Mapper/ParentDepthLimitationMapper.php:
```diff
@@ -29,4 +32,9 @@ protected function getSelectionChoices()
+
+    public function mapLimitationValue(Limitation $limitation)
+    {
+        return $limitation->limitationValues;
```
Minor: This is the same as for the NullLimitationMapper. We could have a `LimitationValueMapper` default implementation which implements the method in this form. But I saw no other cases, and code duplication is minimal.

glye commented on lib/Limitation/LimitationValueMapperInterface.php:
```diff
@@ -0,0 +1,19 @@
+interface LimitationValueMapperInterface
+{
+    /**
+     * Map limitation value to human readable format.
```
Doc could be better here. In some cases an array of strings is returned, that is indeed human readable. But in other cases it's an array of Section objects, or ContentType, or Language, and those are not human readable per se, until they get rendered further down the line.

adamwojs commented:
PR update: 

* Added support for searching the limitation value block definition in multiple resources (this is important from perspective of custom `LimitationType`)
* Added support for  `template` parameter in `ez_render_limitation_value`, which allows to inline override of template used for rendering
* Limitation types without registerd value mapper are now rendered using "fallback block" `ez_limitation_value_fallback`
* Renamed `LimitationRendererInterface` to `LimitationBlockRendererInterface` for naming consistency with `FieldBlockRendererInterface`
* Fixed most of nitpicks

adamwojs commented:
PR ready for final code review.

lserwatka commented:
@adamwojs & @DominikaK could you make a followup PR on the missing documentation?

lserwatka commented:
@adamwojs please merge up to master.
